libevmasm/EVMAssemblyStack.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libevmasm/EVMAssemblyStack.h>
+
+#include <libsolutil/JSON.h>
+#include <liblangutil/Exceptions.h>
+#include <libsolidity/codegen/CompilerContext.h>
+
+#include <tuple>
+
+using namespace solidity::util;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace solidity::evmasm
+{
+
+void EVMAssemblyStack::parseAndAnalyze(std::string const& _sourceName, std::string const& _source)
+{
+	solAssert(!m_evmAssembly);
+	m_name = _sourceName;
+	Json::Value assemblyJson;
+	solRequire(jsonParseStrict(_source, assemblyJson), AssemblyImportException, "Could not parse JSON file.");
+	std::tie(m_evmAssembly, m_sourceList) = evmasm::Assembly::fromJSON(assemblyJson);
+	solRequire(m_evmAssembly != nullptr, AssemblyImportException, "Could not create evm assembly object.");
+}
+
+void EVMAssemblyStack::assemble()
+{
+	solAssert(m_evmAssembly);
+	solAssert(m_evmAssembly->isCreation());
+	solAssert(!m_evmRuntimeAssembly);
+
+	m_object = m_evmAssembly->assemble();
+	m_sourceMapping = AssemblyItem::computeSourceMapping(m_evmAssembly->items(), sourceIndices());
+	if (m_evmAssembly->numSubs() > 0)
+	{
+		m_evmRuntimeAssembly = std::make_shared<evmasm::Assembly>(m_evmAssembly->sub(0));
+		solAssert(m_evmRuntimeAssembly && !m_evmRuntimeAssembly->isCreation());
+		m_runtimeSourceMapping = AssemblyItem::computeSourceMapping(m_evmRuntimeAssembly->items(), sourceIndices());
+		m_runtimeObject = m_evmRuntimeAssembly->assemble();
+	}
+}
+
+LinkerObject const& EVMAssemblyStack::object(std::string const& _contractName) const
+{
+	solAssert(_contractName == m_name);
+	return m_object;
+}
+
+LinkerObject const& EVMAssemblyStack::runtimeObject(std::string const& _contractName) const
+{
+	solAssert(_contractName == m_name);
+	return m_runtimeObject;
+}
+
+std::map<std::string, unsigned> EVMAssemblyStack::sourceIndices() const
+{
+	solAssert(m_evmAssembly);
+	std::map<std::string, unsigned> indices;
+	unsigned index = 0;
+	for (auto const& s: m_sourceList)
+		if (s != CompilerContext::yulUtilityFileName())
+			indices[s] = index++;
+
+	if (indices.find(CompilerContext::yulUtilityFileName()) == indices.end())
+		indices[CompilerContext::yulUtilityFileName()] = index++;
```

Comment:
This adds the utility file name to the source list, even if the assembly was not built from it. Seems wrong. Probably related to the leftovers of the `CompilerStack` hack. I think it should be removed.

Reply:
Also, this way the indices will not match the array returned by `sourceNames()`.

Reply:
Yeah, maybe this function shouldn't rely on the `m_sourceList` vector returned by `evmasm::Assembly::fromJSON()`. Many parts seem to assume the vector indices as source indices, and I think we should avoid that. And probably just having a map from `source_ids` to `source_names` instead. So I'm thinking of a small refactor around `fromJSON` to attempt to fix all these issues related to the indexes.

Reply:
Oh, but we don't have the indices at that stage. Now I think I understand why Alex added that `index` field to the assembly JSON: https://github.com/ethereum/solidity/pull/13673#discussion_r1352582516

Reply:
But the `index` field was for subassemblies and here we're talking about the order of source files. These two should not be related in any way.

But in any case, after the JSON is processed the assumption is that both subassemblies in `m_subs` and the source file names in `m_sourceList` are ordered correctly. The position in the vector is the index.